File: ident-uuid/src/main/scala/net/nicktelford/ident/uuid/Variant.scala
package net.nicktelford.ident.uuid

/** Base trait for UUID variants. */
sealed trait Variant {

  /** Bit-mask for the Variant. */
  def mask: Byte

  /** Value for the Variant. */
  def value: Byte
}

object Variant {

  /** Variant reserved for NCS backward compatibility */
  case object NCS extends Variant {
    val mask: Byte = 0x8
    val value: Byte = 0x0
  }

  /** Variant for UUIDs that adhere to RFC-4122.
   *
   * @see http://www.ietf.org/rfc/rfc4122.txt
   */
  case object RFC extends Variant {
    val mask: Byte = 0xC
    val value: Byte = 0x8
  }

  /** Variant reserved for Microsoft backwards compatibility. */
  case object Microsoft extends Variant {
    val mask: Byte = 0xE
    val value: Byte = 0xC
  }

  /** Variant for UUIDs that adhere to a custom scheme. */
  case object Custom extends Variant {
    val mask: Byte = 0xE
    val value: Byte = 0xE
  }

  /** Special Variant only to be used by the NilUUID. */
  case object Nil extends Variant {
    val mask: Byte = 0x0
    val value: Byte = 0x0
  }
}

File: ident-java/src/main/java/net/nicktelford/ident/Variant.java
package net.nicktelford.ident;

/**
 * Enumeration of UUID Variants.
 */
public enum Variant {

    /**
     * Variant reserved for NCS backward compatibility.
     */
    NCS (0x8, 0x0),

    /**
     * Variant for UUIDs that adhere to RFC-4122.
     *
     * @see http://www.ietf.org/rfc/rfc4122.txt
     */
    RFC (0xC, 0x8),

    /**
     * Variant reserved for Microsoft backwards compatibility.
     */
    Microsoft (0xE, 0xC),

    /**
     * Variant for UUIDs that adhere to a custom scheme.
     * <p/>
     * <i>Note: {@link UUID}s with a {@code Custom} {@link Variant} are not
     * guaranteed to be universally unique as custom UUID formats will vary
     * between vendors.</i>
     */
    Custom (0xE, 0xE),

    /**
     * Special Variant only to be used by the {@link UUID#Nil Nil UUID}.
     */
    Nil (0x0, 0x0);

    private final byte mask;
    private final byte value;

    Variant(final int mask, final int value) {
        this((byte) mask, (byte) value);
    }

    Variant(final byte mask, final byte value) {
        this.mask = mask;
        this.value = value;
    }

    /**
     * The {@link Variant} bit-mask, used to determine which bits encode the
     * {@link #value()} in a {@link UUID}.
     *
     * @return the bit-mask of this {@link Variant }that determines the bits
     *         that encode the {@link #value()}.
     */
    public byte mask() {
        return mask;
    }

    /**
     * The {@link Variant} value, which uniquely describes each {@link Variant}.
     *
     * @return the {@link Variant} value.
     */
    public byte value() {
        return value;
    }
}
